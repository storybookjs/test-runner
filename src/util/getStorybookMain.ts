import { join, resolve } from 'path';
import { serverRequire } from 'storybook/internal/common';
import type { StorybookConfig } from 'storybook/internal/types';
import dedent from 'ts-dedent';

export const storybookMainConfig = new Map<string, StorybookConfig>();

export const getStorybookMain = (configDir = '.storybook') => {
  if (storybookMainConfig.has(configDir)) {
    return storybookMainConfig.get(configDir) as StorybookConfig;
  } else {
    storybookMainConfig.set(configDir, serverRequire(join(resolve(configDir), 'main')));
  }

  const mainConfig = storybookMainConfig.get(configDir);

  if (!mainConfig) {
    throw new Error(
      `Could not load main.js in ${configDir}. Is the "${configDir}" config directory correct? You can change it by using --config-dir <path-to-dir>`
    );
  }

  if (!mainConfig.stories || mainConfig.stories.length === 0) {
    throw new Error(
      dedent`
        Could not find stories in main.js in "${configDir}".
        If you are using a mono-repository, please run the test-runner only against your sub-package, which contains a .storybook folder with "stories" defined in main.js.
        You can change the config directory by using --config-dir <path-to-dir>
        `
    );
  }

  return mainConfig;
};

export function resetStorybookMainCache() {
  storybookMainConfig.clear();
}
