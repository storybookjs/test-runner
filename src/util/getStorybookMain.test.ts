import { getStorybookMain, resetStorybookMainCache, storybookMainConfig } from './getStorybookMain';
import * as coreCommon from 'storybook/internal/common';

jest.mock('storybook/internal/common');

describe('getStorybookMain', () => {
  beforeEach(() => {
    resetStorybookMainCache();
  });

  it('should throw an error if no configuration is found', () => {
    expect(() => getStorybookMain('.storybook')).toThrowErrorMatchingSnapshot();
  });

  describe('no stories', () => {
    it('should throw an error if no stories are defined', () => {
      jest.spyOn(coreCommon, 'serverRequire').mockImplementation(() => ({}));

      expect(() => getStorybookMain('.storybook')).toThrowErrorMatchingSnapshot();
    });

    it('should throw an error if stories list is empty', () => {
      jest.spyOn(coreCommon, 'serverRequire').mockImplementation(() => ({ stories: [] }));

      expect(() => getStorybookMain('.storybook')).toThrowErrorMatchingSnapshot();
    });
  });

  it('should return mainjs', () => {
    const mockedMain = {
      stories: [
        {
          directory: '../stories/basic',
          titlePrefix: 'Example',
        },
      ],
    };

    jest.spyOn(coreCommon, 'serverRequire').mockImplementation(() => mockedMain);

    const res = getStorybookMain('.storybook');
    expect(res).toMatchObject(mockedMain);
  });

  it('should return the configDir value if it exists', () => {
    const mockedMain = {
      stories: [
        {
          directory: '../stories/basic',
          titlePrefix: 'Example',
        },
      ],
    };
    storybookMainConfig.set('configDir', mockedMain);

    const res = getStorybookMain('.storybook');
    expect(res).toMatchObject(mockedMain);
  });
});
