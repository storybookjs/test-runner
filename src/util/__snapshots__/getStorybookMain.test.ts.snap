// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`getStorybookMain no stories should throw an error if no stories are defined 1`] = `
"Could not find stories in main.js in ".storybook".
If you are using a mono-repository, please run the test-runner only against your sub-package, which contains a .storybook folder with "stories" defined in main.js.
You can change the config directory by using --config-dir <path-to-dir>"
`;

exports[`getStorybookMain no stories should throw an error if stories list is empty 1`] = `
"Could not find stories in main.js in ".storybook".
If you are using a mono-repository, please run the test-runner only against your sub-package, which contains a .storybook folder with "stories" defined in main.js.
You can change the config directory by using --config-dir <path-to-dir>"
`;

exports[`getStorybookMain should throw an error if no configuration is found 1`] = `"Module not found"`;
